fml/message_loop.cc

```diff
@@ -0,0 +1,59 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/message_loop.h"
+#include <utility>
+#include "flutter/fml/message_loop_impl.h"
+#include "flutter/fml/task_runner.h"
+#include "flutter/fml/thread_local.h"
+#include "lib/ftl/memory/ref_counted.h"
+#include "lib/ftl/memory/ref_ptr.h"
+
+namespace fml {
+
+FML_THREAD_LOCAL ThreadLocal TLSMessageLoop([](intptr_t value) {
+  delete reinterpret_cast<MessageLoop*>(value);
+});
+
+MessageLoop& MessageLoop::GetCurrent() {
+  auto loop = reinterpret_cast<MessageLoop*>(TLSMessageLoop.Get());
+  if (loop != nullptr) {
+    return *loop;
+  }
+
+  loop = new MessageLoop();
```

Comment:
Under the current scheme, the message loop is guaranteed to be initialized on the thread it will run. This is important because the platform specific message loop implementations are initialized at the same time as the `fml::MessageLoop`.  That gives the implementations access to the correct platform message loop  (`CFRunLoopRef` in this case) so it can add whatever sources or timers it needs to add to service the `fml::MessageLoop` (See `MessageLoopDarwin::MessageLoopDarwin()`). We cant depend on doing this lazily in `MessageLoop::Run` because we may not actually get a chance to run the message loop. This is the case with the default message loop on iOS and Mac (and also on Android).

As written, message loop is ready for use without further configuration. Furthermore, if the platform has already decided to run the message loop for us, we don't care because our sources and timers have already been configured. I find this way a lot cleaner than the base implementation where there had to be a fake `Run` call made on the message loop associated with the main thread (since it was already running). Also, the configuration will be thread sensitive so I'd prefer to do it on my own.